R/stats.R

```diff
@@ -170,12 +153,119 @@ cohens_d <- function(object, group = group_col(object),
                     stringsAsFactors = FALSE)
   }
 
-  rownames(ds) <- ds$Feature_ID
+  if (is.null(time_levels)) {
+    colnames(ds)[2] <- paste0("Cohen_d_", group_levels[1], "_", group_levels[2])
+  } else {
+    colnames(ds)[2] <- paste0("Cohen_d_", group_levels[1], "_", group_levels[2],
+                              "_", time_levels[2], "_minus_", time_levels[1])
+  }
 
   log_text("Cohen's D computed.")
   ds
 }
 
+
+#' Cohen's D
+#'
+#' Computes Cohen's D for each feature. If time and ID are supplied,
+#' change between two time points is computed for each subject,
+#' and Cohen's d is computed from the changes
+#'
+#' @param object a MetaboSet object
+#' @param id character, name of the subject ID column
+#' @param group character, name of the group column
+#' @param time character, name of the time column
+#'
+#' @return data frame with Cohen's d for each feature
+#'
+#' @examples
+#' d_results <- cohens_d(drop_qcs(example_set))
+#' d_results_time <- cohens_d(drop_qcs(example_set),
+#'                            time = "Time", id = "Subject_ID")
+#'
+#' @export
+cohens_d <- function(object, group = group_col(object),
+                     id = NULL, time = NULL) {
+  res <- NULL
+  group_combos <- combn(levels(pData(object)[, group]), 2)
+
+  count_obs_geq_than <- function(x, n) {
+    sum(x >= n)
+  }
+
+  if(is.null(time)) {
+    for (i in seq_len(ncol(group_combos))) {
+      object_split <- object[, which(
+        pData(object)[, group] %in% c(group_combos[1, i], group_combos[2, i])
+      )]
+      pData(object_split) <- droplevels(pData(object_split))
+
+      if (is.null(res)) {
+        res <- cohens_d_fun(object_split, group, id, time)
+        } else {
+        res <- dplyr::full_join(res,
+                                cohens_d_fun(object_split, group, id, time),
+                                by = "Feature_ID"
+        )
+      }
+    }
+  } else {
+    if (is.null(id)) {
+      stop("Please specify id column.", call. = FALSE)
+    }
+    time_combos <- combn(levels(pData(object)[, time]), 2)
+    for (i in seq_len(ncol(group_combos))) {
+      for (j in seq_len(ncol(time_combos))) {
+        object_split <- object[, which(
+          pData(object)[, group] %in% c(group_combos[1, i], group_combos[2, i]) &
+            pData(object)[, time] %in% c(time_combos[1, j], time_combos[2, j])

```

Comment:
A very minor suggestion to make the code cleaner: (I'm pretty sure this works)

```suggestion
          pData(object)[, group] %in% group_combos[, i]) &
          pData(object)[, time] %in% time_combos[, j])
```